src/main/java/com/gregtechceu/gtceu/common/item/ColorSprayBehaviour.java

```diff
@@ -194,17 +195,52 @@ public InteractionResult useOn(UseOnContext context) {
         var facing = context.getClickedFace();
         var pos = context.getClickedPos();
         var stack = context.getItemInHand();
+        var block = level.getBlockState(pos).getBlock();
+
+        int maxBlocksToRecolor = Math.max(1, player != null && player.isCrouching() ? ConfigHolder.INSTANCE.tools.sprayCanChainLength : 1);
+
         if (player != null) {
-            if (!tryPaintBlock(player, level, pos, facing)) {
-                return InteractionResult.PASS;
+            for (int i = 0; i < maxBlocksToRecolor; i++) {
+
+                // Stop once we hit another block to prevent accidentally recoloring stuff
+                if (level.getBlockState(pos).getBlock() != block) {
+                    break;
+                }
+
+                if (!tryPaintBlock(player, level, pos, facing)) {
+                    return InteractionResult.PASS;
+                }
+
+                if (!useItemDurability(player, context.getHand(), stack, empty.get())) {
+                    break;
+                }
+
+                switch (getPaintDirection(player)) {
+                    case UP -> pos = pos.offset(0, 1, 0);
+                    case DOWN -> pos = pos.offset(0, -1, 0);
+                    case NORTH -> pos = pos.offset(0, 0, -1);
+                    case SOUTH -> pos = pos.offset(0, 0, 1);
+                    case WEST -> pos = pos.offset(-1, 0, 0);
+                    case EAST -> pos = pos.offset(1, 0, 0);
```

Comment:
`pos` could be a `MutableBlockPos` so you don't have to create a new one every time.
then you could also just do `pos.move(getPaintDirection(player))`.